homeassistant/components/control4/config_flow.py

```diff
@@ -0,0 +1,205 @@
+    }
+)
+
+
+@callback
+def configured_instances(hass):
+    """Return a set of configured Control4 instances."""
+    return {
+        entry.data.get(CONF_CONTROLLER_NAME)
+        for entry in hass.config_entries.async_entries(DOMAIN)
+    }
+
+
+class Control4Validator:
+    """Validates that config details can be used to authenticate and communicate with Control4."""
+
+    def __init__(self, host, username, password):
+        """Initialize."""
+        self.host = host
+        self.username = username
+        self.password = password
+        self.account = C4Account(self.username, self.password)
+        self.controller_name = None
+        self.director_bearer_token = None
+        self.director = None
+
+    async def authenticate(self) -> bool:
+        """Test if we can authenticate with the Control4 account API."""
+        try:
+            # Authenticate with Control4 account
+            await self.account.getAccountBearerToken()
+
+            # Get controller name
+            account_controllers = await self.account.getAccountControllers()
+            self.controller_name = account_controllers["controllerCommonName"]
+
+            # Get bearer token to communicate with controller locally
+            self.director_bearer_token = await self.account.getDirectorBearerToken(
+                self.controller_name
+            )
+            return True
+        except (BadCredentials, Unauthorized) as exception:
+            _LOGGER.error(exception)
+            return False
+
+    async def connect_to_director(self) -> bool:
+        """Test if we can connect to the local Control4 Director."""
+        try:
+            self.director = C4Director(self.host, self.director_bearer_token)
+            await self.director.getAllItemInfo()
+            return True
+        except Exception as exception:  # pylint: disable=broad-except
+            _LOGGER.error(exception)
+            return False
+
+
+async def validate_input(hass: core.HomeAssistant, data):
+    """Validate the user input allows us to connect.
+
+    Data has the keys from DATA_SCHEMA with values provided by the user.
+    """
+
+    hub = Control4Validator(data["host"], data["username"], data["password"])
+
+    if not await hub.authenticate():
+        raise InvalidAuth
+
+    if not await hub.connect_to_director():
+        raise CannotConnect
+
+    # If you cannot connect:
+    # throw CannotConnect
+    # If the authentication is wrong:
+    # InvalidAuth
+
+    # Return info that you want to store in the config entry.
+    return {
+        "controller_name": hub.controller_name,
+    }
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Control4."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_POLL
+
+    async def async_step_user(self, user_input=None):
+        """Handle the initial step."""
+        errors = {}
+        if user_input is not None:
+
```

Comment:
Does the controller have a unique id?

```
await self.async_set_unique_id(info.serial_number)
self._abort_if_unique_id_configured()
```

https://developers.home-assistant.io/docs/entity_registry_index/#unique-id-requirements

Reply:
Yep, `controller_name` is actually unique and cannot be changed by the user (I'm pretty sure - Control4 only allows certified dealers to access controller config, so I can't test this)

Reply:
If the dealer can change it, it is not exactly stable.  Is there something better we can use?

Reply:
Unfortunately there isn't - this is the way the Control4 API/apps uniquely identify controllers, so I think that's the best way to do it.

Reply:
We can't use something that we configure as it doesn't meet the unique id requirements.

Its better to not set a unique id in this case and check to see if an existing entry has the same controller name instead.

`for entry in self._async_current_entries()`

Reply:
It looks like the mac address is provided here ?
https://github.com/lawtancool/pyControl4/blob/master/pyControl4/account.py#L124

Reply:
Yep, that `"controllerCommonName": "control4_MODEL_MACADDRESS",` string is what I'm using as the unique id. I'm pretty sure this is unique, it's just that the Control4 API is not publicly documented so I can't confirm.

Reply:
What does that string actually look like in production ?

Reply:
It looks something like `control4_hc800_000FFF500000`

Reply:
Is the mac address the actual mac address of the device on your network ?

Reply:
Yes it is.

Reply:
Perfect,

```

from homeassistant.helpers.device_registry import format_mac


control4, model, mac = controller_common_name.split('_', 3)
formatted_mac = format_mac(mac)
```


Then use the `formatted_mac` for the uniqueid